server/src/main/java/org/opensearch/cluster/routing/allocation/decider/RemoteStoreAllocationDecider.java

```diff
@@ -0,0 +1,110 @@
+import org.opensearch.cluster.routing.ShardRouting;
+import org.opensearch.cluster.routing.allocation.RoutingAllocation;
+import org.opensearch.common.settings.ClusterSettings;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.node.remotestore.RemoteStoreNodeService;
+import org.opensearch.node.remotestore.RemoteStoreNodeService.Direction;
+import org.opensearch.cluster.node.DiscoveryNode;
+
+/**
+ * An allocation decider to oversee shard allocation or relocation to remote-store enabled
+ * nodes. If Direction is set as "REMOTE_STORE", new primary shards can only go to remote-store enabled
+ * nodes. Replica shards can go to remote nodes only if corresponding primary also exists on a remote node.
+ *
+ * @opensearch.internal
+ */
+public class RemoteStoreAllocationDecider extends AllocationDecider {
+    public static final String NAME = "remote_store_enabled_cluster";
+
+    private Direction direction;
+
+    public RemoteStoreAllocationDecider(Settings settings, ClusterSettings clusterSettings) {
+        this.direction = RemoteStoreNodeService.DIRECTION_SETTING.get(settings);
+        clusterSettings.addSettingsUpdateConsumer(RemoteStoreNodeService.DIRECTION_SETTING, this::setDirection);
+    }
+
+    private void setDirection (Direction direction) {
+        this.direction = direction;
+    }
+
+    @Override
+    public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        DiscoveryNode targetNode = node.node();
+        if (!shardRouting.assignedToNode()){
+            if (shardRouting.primary()) {
+                if (direction.equals(Direction.REMOTE_STORE) && !targetNode.isRemoteStoreNode()) {
+                    return allocation.decision(Decision.NO, NAME,
+                        "for REMOTE_STORE direction, new primary shards can not be allocated to non-remote nodes");
+                }
+            }
+            else {
+                ShardRouting primaryShardRouting = allocation.routingNodes().activePrimary(shardRouting.shardId());
+                if (primaryShardRouting == null) {
+                    // ReplicaAfterPrimaryActiveAllocationDecider should prevent this case from occurring
+                    return allocation.decision(Decision.NO, NAME, "no active primary shard yet");
+                }
+
+                DiscoveryNode primaryShardNode = allocation.routingNodes()
+                    .stream()
+                    .filter(nd -> nd.nodeId().equals(primaryShardRouting.currentNodeId()))
+                    .findFirst().get().node();
+
+                if (direction.equals(Direction.REMOTE_STORE)) {
+                    if (!primaryShardNode.isRemoteStoreNode() && targetNode.isRemoteStoreNode()) {
+                        return allocation.decision(Decision.NO, NAME,
+                            "can not allocate replica shard on a remote node when primary shard is not already active on some remote node");
```

Comment:
nit: Can we change the message to something like below:

```
cannot allocate replica shard copy on a remote node since primary shard copy is not yet migrated to remote
```